### PR TITLE
PLATFORM-386: update-iam-role-for-controlplane-eks

### PR DESCRIPTION
**Pull request by thameezb**
- creates new PB policy distinct to old BYOC assumed role
- adds new Controlplane policy to DC BYOC assumed role
- creates three new EKS related PBs

Have tested that it works when creating, modifying and deleting new KRuntime Clusters in BYOC networks.

### CODE REVIEW

**Review by eposinitskiy — DISMISSED** [review]

**Review by mjwnkuna — DISMISSED** [review]

File: iam.tf [hunk 1]
```diff
@@ -318,17 +488,586 @@ data "aws_iam_policy_document" "doublecloud_permissions" {
...
+      "iam:GetRole",
+      "iam:ListAttachedRolePolicies",
+    ]
+    resources = ["arn:aws:iam::${local.account_id}:role/DoubleCloud/*"]
+  }
+
+  statement {
+    sid       = "EKSAllowCreateServiceLinkedRole"
+    effect    = "Allow"
+    resources = ["*"]
```
Comment (denisovval) [inline iam.tf:664]: We allow the creation of any service-linked role that can lead to misuse. Can we add some filters here? Maybe add tag-based restrictions?

File: iam.tf [hunk 2]
```diff
@@ -34,8 +46,163 @@ data "aws_iam_policy_document" "trusted_policy" {
...
+      "ec2:CreateVpc",
+      "ec2:DeleteVpc",
+    ]
+    resources = ["*"]
+  }
+
+  statement {
+    effect    = "Allow"
+    actions   = ["ec2:*"]
+    resources = ["*"]
```
[the remark this answers is not in the document]
Comment (denisovval) [reply iam.tf:93]: Like in other parts where we use VPC arn for example.

File: iam.tf [hunk 3]
```diff
@@ -318,17 +488,586 @@ data "aws_iam_policy_document" "doublecloud_permissions" {
...
+    sid    = "AmazonEBSCSIDriverPolicyEC2CreateDescribeV2"
+    effect = "Allow"
+    actions = [
+      "ec2:CreateSnapshot",
+      "ec2:AttachVolume",
+      "ec2:DetachVolume",
+      "ec2:ModifyVolume",
+      "ec2:Describe*",
+    ]
+    resources = ["*"]
```
Comment (denisovval) [inline iam.tf:968]: This is not limited by permission boundaries which is required as it's allows AttachVolume and CreateSnampshot.

**Review by denisovval — CHANGES_REQUESTED** [review]

File: iam.tf (hunk 2 above)
[the remark this answers is not in the document]
Comment (thameezb) [reply iam.tf:93]: This is the PB, not the actual policy (i.e what we can or cannot do). The aim is this:
- PB is generic with specific filters to stop us from doing major bad things even if we wanted to (e.g https://github.com/doublecloud/terraform-aws-doublecloud-byoc/pull/6/files#diff-2bb47ab35b4fcfacd8f871bd1f4bd594f5d1c2a89c0eb861c0e305175a0ef4ecR81-R88)
- Policies attached to our role concretize these abilities (and therefore all EC2 stuff has the restrictions there)


Remember that once created we cannot: alter the PB, nor change the policies attached to the role, therefore from the customer point of view it should be safe
Finally EC2 has a bunch of different API calls, each with very distinct ways to restrict it. This pollutes the PB as it's defined in the actual location of ability (i.e the role). Additionally we quickly hit the policy size limit if we just repeat

File: iam.tf [hunk 4]
```diff
@@ -318,17 +488,586 @@ data "aws_iam_policy_document" "doublecloud_permissions" {
...
+  }
+
+  statement {
+    sid = "EKSAllowPassRolesDoubleCloud"
+    actions = [
+      "iam:PassRole",
+    ]
+    effect    = "Allow"
+    resources = ["arn:aws:iam::${local.account_id}:role/*"]
```
[the remark this answers is not in the document]
Comment (thameezb) [reply iam.tf:643]: AWS upstream policy has no restriction. 

We currently have the resource defined in the same manner for existing EC2PassRole (therefore I copied that)

File: iam.tf (hunk 1 above)
Comment (thameezb) [reply iam.tf:664]: Negative, the condition is below? i.e we can only create ServiceLinkedRoles for the services defined here -> https://github.com/doublecloud/terraform-aws-doublecloud-byoc/pull/6/files#diff-2bb47ab35b4fcfacd8f871bd1f4bd594f5d1c2a89c0eb861c0e305175a0ef4ecR669-R680

This is required as per AWS upstream policy

File: iam.tf (hunk 3 above)
Comment (thameezb) [reply iam.tf:968]: I do not follow? This is the Permission Boundary for the EKS node.

**Review by denisovval — APPROVED** [review]

**Pull request merged**